compression_algorithms/huffmann/huffmann.go

```diff
@@ -0,0 +1,267 @@
+		chars = append(chars, char)
+	}
+
+	for _, char := range chars {
+		heap.Push(pq, &Node{
+			char:   char,
+			freq:   freqMap[char],
+			isLeaf: true,
+		})
+	}
+
+	for pq.Len() > 1 {
+		left := heap.Pop(pq).(*Node)
+		right := heap.Pop(pq).(*Node)
+
+		if left.char > right.char {
+			left, right = right, left
+		}
+
+		merged := &Node{
+			freq:   left.freq + right.freq,
+			left:   left,
+			right:  right,
+			isLeaf: false,
+			char:   min(left.char, right.char),
+		}
+		heap.Push(pq, merged)
+	}
+
+	return heap.Pop(pq).(*Node)
+}
+
+func min(a, b rune) rune {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func generateCodes(root *Node, prefix string, codes map[rune]string) {
+	if root == nil {
+		return
+	}
+
+	if root.isLeaf {
+		// For single character case, use "0" as the code.
+		if prefix == "" {
+			codes[root.char] = "0"
+		} else {
+			codes[root.char] = prefix
+		}
+		return
+	}
+
+	generateCodes(root.left, prefix+"0", codes)
+	generateCodes(root.right, prefix+"1", codes)
+}
+
+type HuffmanCodec struct {
+	root  *Node
+	codes map[rune]string
+}
+
+func newHuffmanCodec(input string) *HuffmanCodec {
+	// Handle empty input.
+	if input == "" {
+		return &HuffmanCodec{
+			root:  nil,
+			codes: make(map[rune]string),
+		}
+	}
+
+	freqMap := make(map[rune]int)
+	for _, char := range input {
+		freqMap[char]++
+	}
+
+	root := buildHuffmanTree(freqMap)
+	codes := make(map[rune]string)
+	generateCodes(root, "", codes)
+
+	return &HuffmanCodec{
+		root:  root,
+		codes: codes,
+	}
+}
+
+func (hc *HuffmanCodec) compress(input string) string {
+	// Handle empty input.
+	if input == "" {
+		return ""
+	}
+
+	// Handle nil codec (created with empty input).
+	if hc.root == nil {
+		return ""
+	}
+
+	compressed := ""
+	for _, char := range input {
+		compressed += hc.codes[char]
+	}
+	return compressed
+}
+
+func (hc *HuffmanCodec) decompress(compressed string) string {
+	// Handle empty input or nil codec.
+	if compressed == "" || hc.root == nil {
+		return ""
+	}
+
+	// Handle single character case.
+	if hc.root.isLeaf {
+		result := ""
+		for range compressed {
+			result += string(hc.root.char)
+		}
+		return result
+	}
+
+	decoded := ""
+	currentNode := hc.root
+
+	for _, bit := range compressed {
+		if bit == '0' {
+			currentNode = currentNode.left
+		} else {
+			currentNode = currentNode.right
+		}
+
+		if currentNode.isLeaf {
+			decoded += string(currentNode.char)
+			currentNode = hc.root
+		}
+	}
+
+	return decoded
+}
+
+func main() {
+	reader := bufio.NewReader(os.Stdin)
+	fmt.Println("Choose an option:")
+	fmt.Println("1. Enter a string")
+	fmt.Println("2. Provide a filename")
+
+	choice, _ := reader.ReadString('\n')
+	choice = strings.TrimSpace(choice)
+
+	var input string
+	var originalSize int
+
+	switch choice {
+	case "1":
+		fmt.Print("Enter the string: ")
+		input, _ = reader.ReadString('\n')
+		input = strings.TrimSpace(input)
+		originalSize = len(input)
+	case "2":
+		fmt.Print("Enter the filename: ")
```

Comment:
relative filepath? Or something more clearer?

Reply:
I updated the prompt to clarify that the user can provide either a relative or absolute path. This change also handles the case where users may accidentally include quotes around the path, which was causing errors when I tried read the file.